### DISCUSSION TITLE
Should JBang default to run with Java 17?

### DISCUSSION THREAD

**Discussion by maxandersen** [jbangdev/jbang#1088] [Ideas] [answered]
Currently jbang defaults to Java 11 if no other java found.

Should we move it up to Java 17 ? 

Reasons for: 
  records and multi line strings are super useful when writing scripts and utilities

Reasons against:
gradle still seem to suck when running under Java 17 so jbang edit experience aint great
many(?) libraries still have challenges with Java 17 due to the enforced locked down jdk.

Because if reasons against i'm leaning on staying on Java 11 but I wish someone could 
convince me otherwise.

Thoughts?

**Comment by quintesse** [comment, marked as answer]
In principle I'd agree with you, and in a perfect world I'd advocate for moving to the latest LTE the moment it comes out.... but unfortunately the world isn't perfect :-)

So, I'd definitely like to see us make Java 17 default at some point, but right now might cause too much issues. Also because the kind of issues that will appear will be ones most people have no experience with and will not know how to fix.

And a second reason for not doing this (yet) is that we've made it too damn easy to just put `//JAVA 17+` in a file :-)

**Comment by quintesse** [comment]
TL;DR: Make Java 17 the default :+1: 

Bit by bit I'm leaning towards making Java 17 the default. Many of the issues we were seeing have been resolved and we are now also able to handle the integrations externally. So the same reason I gave for _not_ doing this, "we've made it too damn easy to just put //JAVA 17+ in a file", can now be used in favor: "we've made it too damn easy to just put `//JAVA 8` in a file" :-)

For JBang itself it would also be nice if we could go to a newer version, but we need to take into account the possibility that we might want to create a JBang "library" at some point.

We'd have to decide if it's important to have that library generically available for everybody using any JDK that's still being used a lot or if we can simply say that to use the library you have to be on 17+ as well.

Personally I'd say that anyone using the JBang library would do so to interoperate with JBang, not because the code inside JBang would be super useful in some completely unrelated project. In that case it wouldn't be strange to say that you have to be on the same Java version as JBang itself.

PS: but let's not forget that the (small amount of) code needed for running the external integrations would need to be kept on Java 8!
